### DISCUSSION TITLE
Skip test based on a parameter

### DISCUSSION THREAD

**Discussion by WernerFS** [cocotb/cocotb#4288] [Q&A] [answered]
Hello. First I want to thank all developers of cocotb. It is an incredible tool. I have been using it daily for the past months and it is soo good.

I am running some tests that couple with a software simulator. There are some cases of the software simulator that are of no particular interest to me. But given that I cannot tell which are these because they are randomly generated I wanted to know if it could be possible to skip a test based on a parameter. 

Currently my code looks like this:

```Python
@cocotb.test()
@cocotb.parametrize(
    seed=[random.randint(1, 2**32) for _ in range(3000)]
)
async def run_test(dut, seed):
    # Define the error rates for each ErrorType
    simulator_data = Simulator(seed)
    if (not simulator_data.errors): # If no errors nothing to check on hardware
        return TestSuccess()
    run_hdl_simulation()
````

I could run the `Simulator` outside, save it to a file, and then feed only the valid test cases. Still, I wanted to check if it could be possible to set the `cocotb.parameterized` and `cocotb.test` values for each iteration.

**Comment by ktbarrett** [comment, marked as answer]
So you don't need to return `TestSuccess`. That's actually meant to be `raise`'d at any point in the test. If you are running the top-level Task, you can simply `return`.

And I would try this...
```python

valid_test_seeds = []
while len(valid_test_seeds) < 3000:
    seed = random.randint(1, 2**32)
    simulator_data = Simulator(seed)
    if simulator_data.errors:
        valid_test_seeds.append(seed)

@cocotb.test()
@cocotb.parametrize(
    seed=valid_test_seeds,
)
async def run_test(dut, seed):
    run_hdl_simulation()
```

**Reply by WernerFS** [reply]
I get the concept but sadly it doesn't work. The decorator is always evaluated before the code is executed. So valid_test_seeds is not defined. I have been trying to get around this issue but it will require another custom decorator. Right now I am experimenting with `pytest_generate_tests` given that it would create automatic fixtures as decorators around the test function. I honestly feel a bit lost. But will keep updating this if anyone comes to a similar issue.

**Reply by WernerFS** [reply]
It seems that what I want to accomplish looks a lot like `TestFactory`.  Which seems like the only way of passing arguments to test functions. You might want to reconsider deprecating it as it seems to be the `pytest_generate_test` equivalent. Being the only way to dynamically generate tests without having to add another decorator around the `parametrize` decorator.

**Reply by WernerFS** [reply]
Sorry for the nonsense messages. Your solution worked fine in a toy example I tested.
